### ISSUE TITLE
Two questions about the EFT method

### ISSUE THREAD

**Issue by eeeeeei** [vkverma01/EFT#2]
Hi, I have some doubts about the implementation：
1. I couldn't find the transformation of the fully connected layer, it seems an ordinary fully connected layer. Did I miss something?
2. About the margin loss, i.e., the second term of training loss, why it is implemented as L2 norm instead of KL divergence as mentioned in the paper？
Hopes for your reply, Thanks!

**Comment by vkverma01** [comment]
Thanks for your interest, please find the clarification below:

1: We have applied the FC layer transformation where FC layer contains a significant amount of parameters like AlexNet on ImageNet (AlexNet contains total 60M parameters, while FC layer contains 55M parameters. 
It can be applied for the other architecture as follows but it significantly does not save the local parameter. It can be applied as follows: 
The FC layer in the diagonal matrix can be defined as:

A: self.embed=nn.Embedding(1,output_dim) # it contains a vector of output dimension.

Then multiply the output by this embedding layer (elementwise).

B: out = self.linear(feat)*self.embed(torch.cuda.LongTensor([0]))
Then freeze the linear layer.


2: In the KLD between the two multivariate normals if we assume that \Sigma_1=\Sigma2=I then it reduces to L2 Norm. 
We have tried both and later find that L2 norm is the simplest and give the same result. Directly it can be replaced with the torch.nn.functional.kl_div(input, target) to get the actual KDL loss.

Hope this helps. 

Thanks

**Comment by eeeeeei** [comment]
> Thanks for your interest, please find the clarification below:
> 
> 1: We have applied the FC layer transformation where FC layer contains a significant amount of parameters like AlexNet on ImageNet (AlexNet contains total 60M parameters, while FC layer contains 55M parameters.
> It can be applied for the other architecture as follows but it significantly does not save the local parameter. It can be applied as follows:
> The FC layer in the diagonal matrix can be defined as:
> 
> A: self.embed=nn.Embedding(1,output_dim) # it contains a vector of output dimension.
> 
> Then multiply the output by this embedding layer (elementwise).
> 
> B: out = self.linear(feat)*self.embed(torch.cuda.LongTensor([0]))
> Then freeze the linear layer.
> 
> 2: In the KLD between the two multivariate normals if we assume that \Sigma_1==\Sigma2 then it reduces to L2 Norm.
> We have tried both and later find that L2 norm is the simplest and give the same result. Directly it can be replaced with the torch.nn.functional.kl_div(input, target) to get the actual KDL loss.
> 
> Hope this helps.
> 
> Thanks

Thanks for your raply! I have got it!

**Comment by pzSuen** [comment]
> 

Hello, 

I am puzzled about what you say `It can be applied for the other architecture as follows but it significantly does not save the local parameter`. Can you explain this more in detail?

And I notice you said, `We observe that architectures with more parameters in fully connected layers (especially AlexNet) have a
harder time adapting features with just diagonal matrices`  in the last paragraph of the supplementary. Does this mean It's tough to use your method(h=Ef) with only a fully-connected layer(without convolutional layer) to do task-incremental learning?

**Comment by appledora** [comment]
Hi @vkverma01 , I wanted some clarification on why we have to freeze the linear layer after multiplying with the diagonal adapter layer. 
As you mentioned in the earlier comment: 
> B: out = self.linear(feat)*self.embed(torch.cuda.LongTensor([0]))
Then freeze the linear layer.
